src/H5TSpool.c

```diff
@@ -0,0 +1,386 @@
+                HGOTO_DONE((H5TS_thread_ret_t)1);
+
+        /* If there's a task, invoke it, else we're shutting down */
+        if (NULL != pool->head) {
+            H5TS_pool_task_t *task; /* Task to invoke */
+
+            /* Grab our task */
+            task = pool->head;
+            if (task->next)
+                pool->head = task->next;
+            else
+                pool->head = pool->tail = NULL;
+
+            /* Release the pool's mutex */
+            if (H5_UNLIKELY(H5TS_mutex_unlock(&pool->mutex) < 0))
+                HGOTO_DONE((H5TS_thread_ret_t)1);
+            have_mutex = false;
+
+            /* Invoke function for task */
+            (*task->func)(task->ctx);
+
+            /* Free the task node */
+            free(task);
+        }
+        else {
+            assert(pool->shutdown);
+            break;
+        }
+    }
+
+done:
+    /* Release the pool's mutex, if we're holding it */
+    if (have_mutex)
+        if (H5_UNLIKELY(H5TS_mutex_unlock(&pool->mutex) < 0))
+            ret_value = (H5TS_thread_ret_t)1;
+
+    FUNC_LEAVE_NOAPI_NAMECHECK_ONLY(ret_value)
+} /* end H5TS__pool_do() */
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS_pool_create
+ *
+ * Purpose:     Create a new thread pool, with the given # of threads
+ *
+ * Return:      Non-negative on success / Negative on failure
+ *
+ *--------------------------------------------------------------------------
+ */
+herr_t
+H5TS_pool_create(H5TS_pool_t **pool, unsigned num_threads)
+{
+    H5TS_pool_t *new_pool = NULL; /* Newly created pool */
+    unsigned     u;               /* Local index variable */
+    herr_t       ret_value = SUCCEED;
+
+    FUNC_ENTER_NOAPI_NAMECHECK_ONLY
+
+    /* Sanity checks */
+    if (H5_UNLIKELY(NULL == pool))
+        HGOTO_DONE(FAIL);
+    if (H5_UNLIKELY(0 == num_threads))
+        HGOTO_DONE(FAIL);
+
+    /* Allocate new pool */
+    if (H5_UNLIKELY(NULL == (new_pool = H5FL_MALLOC(H5TS_pool_t))))
+        HGOTO_DONE(FAIL);
+
+    /* Initialize pool fields to defaults */
+    memset(new_pool, 0, sizeof(*new_pool));
+    if (H5_UNLIKELY(H5TS_mutex_init(&new_pool->mutex, H5TS_MUTEX_TYPE_PLAIN) < 0))
+        HGOTO_DONE(FAIL);
+    if (H5_UNLIKELY(H5TS_cond_init(&new_pool->cond) < 0))
+        HGOTO_DONE(FAIL);
+
+    /* Allocate array of threads */
+    if (H5_UNLIKELY(NULL == (new_pool->threads = H5FL_SEQ_MALLOC(H5TS_thread_t, num_threads))))
+        HGOTO_DONE(FAIL);
+
+    /* Set # of threads */
+    new_pool->num_threads = num_threads;
+
+    /* Start worker threads */
+    for (u = 0; u < num_threads; u++)
+        /* Create thread, which immediately starts processing tasks */
+        if (H5_UNLIKELY(H5TS_thread_create(&new_pool->threads[u], H5TS__pool_do, (void *)new_pool) < 0))
+            HGOTO_DONE(FAIL);
+
+    /* Check for error when starting threads */
+    if (u != new_pool->num_threads) {
```

Comment:
Since the above logic skips to the done section when a failure occurs, it looks like the logic here should be in the done section to ensure that the correct number of threads is set when a failure to create a thread occurs.

Reply:
Agree